kernel/work.c

```diff
@@ -0,0 +1,966 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * Second generation work queue implementation
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <wait_q.h>
+#include <spinlock.h>
+#include <errno.h>
+#include <ksched.h>
+#include <sys/printk.h>
+
+/* Lock to protect the internal state of all work items, and
+ * pending_cancels.
+ *
+ * No queue locks may be held when this is taken, but a queue lock may
+ * be taken while this is held.
+ */
+static struct k_spinlock lock;
+
+/* Invoked by work thread */
+static void handle_flush(struct k_work *work)
+{
+	struct z_work_flusher *flusher
+		= CONTAINER_OF(work, struct z_work_flusher, work);
+
+	k_sem_give(&flusher->sem);
+}
+
+static inline void init_flusher(struct z_work_flusher *flusher)
+{
+	k_sem_init(&flusher->sem, 0, 1);
+	k_work_init(&flusher->work, handle_flush);
+}
+
+/* List of pending cancellations. */
+static sys_slist_t pending_cancels;
+
+/* Initialize a canceler record and add it to the list of pending
+ * cancels.
+ *
+ * Invoked with work lock held.
+ *
+ * @param canceler the structure used to notify a waiting process.
+ * @param work the work structure that is to be canceled
+ */
+static inline void init_work_cancel(struct z_work_canceller *canceler,
+				    struct k_work *work)
+{
+	k_sem_init(&canceler->sem, 0, 1);
+	canceler->work = work;
+	sys_slist_append(&pending_cancels, &canceler->node);
+}
+
+/* Complete cancellation of a work item and unlock held lock.
+ *
+ * Invoked with work lock held.
+ *
+ * Invoked from a work queue thread.
+ *
+ * Reschedules.
+ *
+ * @param work the work structre that has completed cancellation
+ */
+static void finalize_cancel_locked(struct k_work *work)
+{
+	struct z_work_canceller *wc, *tmp;
+
+	/* Clear this first, so released high-priority threads don't
+	 * see it when doing things.
+	 */
+	atomic_clear_bit(&work->flags, K_WORK_CANCELING_BIT);
+
+	/* Search for and remove the matching container, and release
+	 * what's waiting for the completion.
+	 */
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&pending_cancels, wc, tmp, node) {
```

Comment:
This loop is digging through an slist looking for a specific item to delete.  Virtually always, that means you should be using a dlist instead, where deletion is O(1).

Reply:
This was wrong, because though `sys_slist_remove()` is O(1) the prev-node parameter was incorrect.  That's been corrected; is this OK now?  It's an idiom used multiple other places.

Reply:
That call is O(1), but you're literally looping over the list right here.  And AFAICT, the only purpose of this list is to store an unordered set of things that never gets looped over or inspected in any way except to remove elements.  Unless there's a really clear reason for why you can't spend the extra word of storage in the node, an slist is just the wrong data structure here.

Reply:
But I have to iterate over the list to identify the items that need to be removed (by checking for canceller records that reference the completed work item), and while doing so it's trivial to delete them as I encounter them.  What does switching to dlist get us?